notebooks/maze_tuto.ipynb

```diff
@@ -0,0 +1,699 @@
+    "            next_y -= 1\n",
+    "        if action == Action.right:\n",
+    "            next_y += 1\n",
+    "        return (\n",
+    "            State(next_x, next_y)\n",
+    "            if self.maze.is_an_empty_cell(next_x, next_y)\n",
+    "            else memory\n",
+    "        )\n",
+    "\n",
+    "    def _get_transition_value(\n",
+    "        self, memory: State, action: Action, next_state: Optional[State] = None\n",
+    "    ) -> Value[float]:\n",
+    "        \"\"\"Get the value (reward or cost) of a transition.\n",
+    "\n",
+    "        Set cost to 1 when moving (energy cost)\n",
+    "        and to 2 when bumping into a wall (damage cost).\n",
+    "\n",
+    "        \"\"\"\n",
+    "        #\n",
+    "        return Value(cost=1 if next_state != memory else 2)\n",
+    "\n",
+    "    def _get_initial_state_(self) -> State:\n",
+    "        \"\"\"Get the initial state.\n",
+    "\n",
+    "        Set the start position as initial state.\n",
+    "\n",
+    "        \"\"\"\n",
+    "        return self.start\n",
+    "\n",
+    "    def _get_goals_(self) -> Space[State]:\n",
+    "        \"\"\"Get the domain goals space (finite or infinite set).\n",
+    "\n",
+    "        Set the end position as goal.\n",
+    "\n",
+    "        \"\"\"\n",
+    "        return ListSpace([self.end])\n",
+    "\n",
+    "    def _is_terminal(self, state: State) -> D.T_agent[bool]:\n",
+    "        \"\"\"Indicate whether a state is terminal.\n",
+    "\n",
+    "        Stop an episode only when goal reached.\n",
+    "\n",
+    "        \"\"\"\n",
+    "        return self._is_goal(state)\n",
+    "\n",
+    "    def _get_action_space_(self) -> Space[Action]:\n",
+    "        \"\"\"Define action space.\"\"\"\n",
+    "        return EnumSpace(Action)\n",
+    "\n",
+    "    def _get_observation_space_(self) -> Space[State]:\n",
+    "        \"\"\"Define observation space.\"\"\"\n",
+    "        return MultiDiscreteSpace([self.maze.height, self.maze.width])\n",
+    "\n",
+    "    def _render_from(self, memory: State, **kwargs: Any) -> Any:\n",
+    "        \"\"\"Render visually the maze.\n",
+    "\n",
+    "        Returns:\n",
+    "            matplotlib figure\n",
+    "\n",
+    "        \"\"\"\n",
+    "        # store used matplotlib subplot and image to only update them afterwards\n",
+    "        self._ax, self._image = self.maze.render(\n",
+    "            current_position=memory,\n",
+    "            goal=self.end,\n",
+    "            ax=self._ax,\n",
+    "            image=self._image,\n",
+    "        )\n",
+    "        return self._image.figure\n",
+    "\n",
+    "    def heuristic(self, s: State) -> Value[float]:\n",
+    "        \"\"\"Heuristic to be used by search algorithms.\n",
+    "\n",
+    "        Here Manhattan distance to goal.\n",
+    "\n",
+    "        \"\"\"\n",
+    "        return Value(cost=sqrt((self.end.x - s.x) ** 2 + (self.end.y - s.y) ** 2))"
```

Comment:
This is usual Euclidean distance, not Manhattan.  BTW do we really need sqrt?

Reply:
Euh you are right of course. I juste copy/paste a comment from Alex without thinking further... my bad